File: implementations/java/Check.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Check {
    public static void main(String[] args) throws IOException {
        List<String[]> mismatches = new ArrayList<>();
        List<String[]> downloadFailures = new ArrayList<>();
        int count = 0;
        String baseUrl = "https://256t.org";

        for (Path path : Files.list(CidUtil.CIDS_DIR).sorted().toList()) {
            if (Files.isDirectory(path)) {
                continue;
            }
            count++;
            String cid = path.getFileName().toString();
            byte[] localContent = CidUtil.readFile(path);
            String expected = CidUtil.computeCid(localContent);
            
            // Check local CID file
            if (!cid.equals(expected)) {
                mismatches.add(new String[]{cid, expected});
            }
            
            // Check downloaded content
            try {
                CidUtil.DownloadResult result = CidUtil.downloadCid(baseUrl, cid);
                if (!result.isValid) {
                    downloadFailures.add(new String[]{cid, result.computed});
                } else if (!Arrays.equals(result.content, localContent)) {
                    downloadFailures.add(new String[]{cid, "content mismatch with local file"});
                }
            } catch (Exception e) {
                downloadFailures.add(new String[]{cid, e.getMessage()});
            }
        }

        boolean hasErrors = false;

        if (!mismatches.isEmpty()) {
            System.out.println("Found CID mismatches:");
            for (String[] mismatch : mismatches) {
                System.out.printf("- %s should be %s%n", mismatch[0], mismatch[1]);
            }
            hasErrors = true;
        }

        if (!downloadFailures.isEmpty()) {
            System.err.println("Found download validation failures:");
            for (String[] failure : downloadFailures) {
                System.err.printf("- %s: %s%n", failure[0], failure[1]);
            }
            hasErrors = true;
        }

        if (hasErrors) {
            System.exit(1);
        }

        System.out.printf("All %d CID files match their contents.%n", count);
        System.out.printf("All %d downloaded CIDs are valid.%n", count);
    }
}

File: implementations/scala/Check.scala
import java.nio.file.Files

object Check {
  def main(args: Array[String]): Unit = {
    val entries = Option(Files.list(Cid.CidsDir)).getOrElse {
      System.err.println(s"Unable to list cids directory: ${Cid.CidsDir}")
      System.exit(1)
      return
    }

    val sorted = try entries.toArray.map(_.asInstanceOf[java.nio.file.Path]).sorted
    finally entries.close()

    var mismatches = scala.collection.mutable.ListBuffer[(String, String)]()
    var downloadFailures = scala.collection.mutable.ListBuffer[(String, String)]()
    var count = 0
    val baseUrl = "https://256t.org"

    sorted.foreach { path =>
      if (!Files.isDirectory(path)) {
        count += 1
        val cid = path.getFileName.toString
        val localContent = Files.readAllBytes(path)
        
        // Check local CID file
        Cid.cidFor(path) match {
          case Left(error) =>
            mismatches += ((cid, error))
          case Right(expected) =>
            if (expected != cid) {
              mismatches += ((cid, expected))
            }
        }
        
        // Check downloaded content
        Cid.downloadCid(baseUrl, cid) match {
          case Left(error) =>
            downloadFailures += ((cid, error))
          case Right(result) =>
            if (!result.isValid) {
              downloadFailures += ((cid, result.computed))
            } else if (!java.util.Arrays.equals(result.content, localContent)) {
              downloadFailures += ((cid, "content mismatch with local file"))
            }
        }
      }
    }

    var hasErrors = false

    if (mismatches.nonEmpty) {
      println("Found CID mismatches:")
      mismatches.foreach { case (cid, expected) =>
        println(s"- $cid should be $expected")
      }
      hasErrors = true
    }

    if (downloadFailures.nonEmpty) {
      System.err.println("Found download validation failures:")
      downloadFailures.foreach { case (cid, error) =>
        System.err.println(s"- $cid: $error")
      }
      hasErrors = true
    }

    if (hasErrors) {
      System.exit(1)
    } else {
      println(s"All $count CID files match their contents.")
      println(s"All $count downloaded CIDs are valid.")
    }
  }
}
